add the initial draft for phoenix with surface (#1)
herminiotorres/blog: 2022/instrumenting-phoenix-with-surface.md

> # Instrumenting Phoenix with Surface

Reviewer: I wouldn't use the work "Instrumenting" here. "Instrumenting" tends to mean adding observability tooling like logs, metrics, or traces. Maybe something like: "Building LiveView Apps with Surface Components" ?

Author: I will be changing the name of the `file` and the `title` too. And I will trying my best of to improve or explain a bit more and wrap up.

Should I spiced up the conclusion to callout to follow for the next parts of this series?

Reviewer: Yes I think mentioning that you plan to publish more articles on this topic in the conclusion is a good idea!

> # Instrumenting Phoenix with Surface

Reviewer: It's really nice when the introduction to a post tells the reader what the post will cover. This gives the reader a good idea of what they will learn and why this post is worth their time. Can you add a sentence or two like:

"In this post, I will tell you what Surface is and why it's so great. Then, you'll learn how to build a simple Surface component and use it in your Phoenix app."

> What is Surface? According to website describe itself as a server-side rendering component library for Phoenix. You can see Surface as a Template DSL to render webpages with a minimal Elixir portion. And why this is good, it because you it be more free to use structures which make more sense and more close to frontend instead of be more functional and close to the language itself. Also, that been said, we can see how to install and use surface with your phoenix project.

Reviewer: ```suggestion
What is Surface? According to the website,  Surface is a server-side rendering component library for Phoenix. You can see Surface as a Template DSL to render webpages with a minimal Elixir portion. Surface allows you to model your front-end components more intuitively. 

Let's start by taking a look at how to install and use Surface in your Phoenix project.
```

Reviewer: I'm not sure what you mean by: `You can see Surface as a Template DSL to render webpages with a minimal Elixir portion`. Could you explain that a bit more?

Author: Yes, sure, I get this from the surface website, but I watch some of the oldest Marlus's talk, and he said the propose to create Surface:
- Create Surface as a Template DSL - Surface has its own template DSL, and compares with Phoenix Template, we are able to see things like `if-elseif-else`, `case`, and some other things.
- Surface use only a small subset of the Language itself instead of Phoenix Templates, Phoenix Template get a full access to the entire Language, and for front-end templates, its only make sense to have this small subset of the Language, and a unique DSL for template where it could be a bit different than the intention of a funcional language.

Reviewer: I would give a brief definition of a "Template DSL" here for those who are unfamiliar and cut the language of "...with only a small Elixir portion". Then I think this statement will make more sense.

> Once, you have your phoenix(`mix phx.new my_app`) project, the next step should be add Surface as a dependency in `mix.exs`:

Reviewer: ```suggestion
Once you have your phoenix(`mix phx.new my_app`) project, the next step is to add Surface as a dependency in your `mix.exs`:
```

> Once, you have your phoenix(`mix phx.new my_app`) project, the next step should be add Surface as a dependency in `mix.exs`:

Reviewer: I would start a new sub-section here: "Installing Surface"

```elixir
def deps do
  [
    {:surface, "~> 0.7.3"}
  ]
end
```

> After fetching the dependencies with `mix do deps.get, deps.compile`, you can run the `surface.init` task to update the necessary files in your project. Besides, before that it's good to run `mix help surface.init` first, and check out all the options it can used.

Reviewer: ```suggestion
After fetching the dependencies with `mix do deps.get, deps.compile`, you can run the `surface.init` task to update the necessary files in your project. But first, run `mix help surface.init` and check out all the options it provides.
```

```shell
mix surface.init
Compiling 14 files (.ex)
Generated my_app app

Note: This task will change existing files in your project.

Make sure you commit your work before running it, especially if this is not a fresh phoenix project.

Do you want to continue? [Yn] y
* patching .formatter.exs
* patching .gitignore
* patching assets/js/app.js
* patching config/config.exs
* patching config/dev.exs
* patching lib/my_app_web.ex
* patching mix.exs

Finished running 10 patches for 7 files.
10 changes applied, 0 skipped.

The following dependencies were updated/added to your project:

  * surface_formatter

Do you want to fetch and install them now? [Yn] y
Resolving Hex dependencies...
```

> Lets checking in each file since surface have been installed and understand one by one.

Reviewer: ```suggestion
Let's checkout each file Surface generated and understand them better, one by one.
```

> Lets checking in each file since surface have been installed and understand one by one.

Reviewer: Since you're taking a pretty deep dive into the generated files/changes that `surface init` brings, I'd start a new subsection here, "Exploring the Generated Files". Then, for each file, show a code snippet that includes the code change that the generator made before explaining that change.

> The `.formatter.exs`, this means we don't need to take care of it how to format our surface code since the [suface_formatter](https://github.com/surface-ui/surface_formatter) can handle for us, so when you need to format your suface code, you just run in the terminal `mix surface.format`, and done, all your surface code it will formatted properly.

Reviewer: ```suggestion
Surface has changed the `.formatter.exs` file so that our Surface code will be formatted automatically  with the help of the [suface_formatter](https://github.com/surface-ui/surface_formatter). Now, when you run `mix surface.format`, all your surface code it will formatted properly.
```

> The `assets/js/app.js`, it will import `Hooks`, and pass with liveSocket. And you can check [here](https://surface-ui.org/js_interop) how to use Hooks with surface, and in case you like to understand how to create and use a Hook with LiveView, you can checking out [here](https://hexdocs.pm/phoenix_live_view/js-interop.html).

Reviewer: ```suggestion
Surface has also updated the `assets/js/app.js` file to import `Hooks`, and pass them to the call to initialize the liveSocket. You can go [here](https://surface-ui.org/js_interop) to learn how to use Hooks with Surface, and [here](https://hexdocs.pm/phoenix_live_view/js-interop.html) to learn how to use Hooks with LiveView.
```

> The `config/config.exs`, it will set a `ErrorTag` form component when your form has some errors and you can see visualy on the page.

Reviewer: ```suggestion
Next up, let's take a look at the `config/config.exs` file. <Then show the code snippet> Here, it sets an `ErrorTag` form component when your form has some errors so that you can see changeset errors for the form visually on the page.
```

> The `config/dev.exs`, it will set a reload compilers, and what does mean? it means every time we change and save our surface component, the compiler it will compiles the newest version of our code, and reload the app in case you can see the changes.

Reviewer: ```suggestion
The Surface generator also updated the `config/dev.exs` file by adding a reload compiler for Surface. This means that every time we change and save our Surface component, the compiler will compiles the newest version of our code and reload the app so that you can see the changes without having to refresh the page in the browser.
```

> The `lib/my_app_web.ex`, it will import the surface for all the views.

Reviewer: ```suggestion
The Surface generator updated the `lib/my_app_web.ex` to import Surface to all of the views.
```

> The `mix.exs`, it will have the surface compiler, for how to compile the surface code, and also raise any erro when its has, and add at least two dependencies, and it was `surface` itself, and `surface_formater`, but it could add the `surface_catalogue`.

Reviewer: ```suggestion
The `mix.exs` now contains the Surface compiler. So, your app now knows how to compile Surface code. That file now also includes two new dependencies--`Surface` itself and `surface_formater`.
```

> So, last not least, let's go and implements our tradicional "hello world" in Surface.

Reviewer: ```suggestion
Now we're ready to implement the traditional "hello world" in Surface.
```

Reviewer: I would introduce the thing you're building here. If you're building a simple counter example, say that.

> We need to create a new route on `lib/my_app_web/router.ex`, and should be called:

Reviewer: Create a new subsection here: "Building Your First Surface Component"

> We need to create a new route on `lib/my_app_web/router.ex`, and should be called:

Reviewer: ```suggestion
We need to create a new route in `lib/my_app_web/router.ex`, like this:
```

```diff
scope "/", MyAppWeb do
  pipe_through :browser

+ live "/count", CounterLive
end
```

Create the `live` folder in `lib/my_app_web/live/`, and then create your `counter_live.ex`.

> And the code inside this file should be like this:

Reviewer: ```suggestion
Your `CounterLive` live view module should look like this:
```

```elixir
defmodule MyAppWeb.CounterLive do
  use Surface.LiveView

  data count, :integer, default: 0

  def render(assigns) do
    ~F"""
    <div>
      <h1 class="title">
        {@count}
      </h1>
      <div>
        <button class="button is-info" :on-click="dec">
          -
        </button>
        <button class="button is-info" :on-click="inc">
          +
        </button>
      </div>
    </div>
    """
  end

  def handle_event("inc", _value, socket) do
    {:noreply, update(socket, :count, &(&1 + 1))}
  end

  def handle_event("dec", _value, socket) do
    {:noreply, update(socket, :count, &(&1 - 1))}
  end
end
```

> Lets see line by line and understand each one of them.

Reviewer: ```suggestion
Let's break this down line by line to understand how Surface works.
```

> The `use Surface.LiveView`, it just a wrapper component around `Phoenix.LiveView`. In LiveView we normally use `use MyAppWeb, :live_view`.

Reviewer: ```suggestion
The `use Surface.LiveView` brings in the Surface behavior. Surface components are really just a wrapper around LiveView components.
```

> The [`data`](https://surface-ui.org/data) macro its the private API for Surface components in general, this means the component doesn't receive any external data when its called, and because this is a live_view itself it make sense has only private API. Also, we could skip to create the [`mount/3`](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html#c:mount/3) function, since we don't doing nothing complext that much to feel necessarily to have a mount function.

Reviewer: From this sentence, its not clear what the `data` macro function does. Could you clarify that is defines what Surface calls "data assigns" and specifies what keys the component expect to be present in socket assigns? And clarify that "data assigns" are data in socket assigns that are created and managed internally in the component, as opposed to some assigns that might get passed in from the parent when the component is rendered.

Author: Sure, I will explain and put on this direction has you suggest, and maybe put some comparison between a regular liveview and surface data assigns.

> The [`render/1`](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html#c:render/1) function it will rendering our template, and instead of using the `HEEX` sigil(`~H`), we re been using the Surface sigil(`~F`). And one of the ergonomics for Surface templates its we can use `{}` for embeded code, and not just only for embeded thing in HTML attributes for tags. And it has `:on-click` handle event for `inc` and `dec`, and to get more context for this handle event directive checking out [here](https://surface-ui.org/events#using-the-:on-[event]-directive).

Reviewer: ```suggestion
The [`render/1`](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html#c:render/1) function renders our template. Since this is a Surface component, instead of using the `HEEX` sigil(`~H`), we use the Surface sigil(`~F`). Let's take a closer look at the markup we're rendered now. You'll notice that we use the `{}` for _all_ embedded code, not just  for embedded code in HTML tags. This is one of the ergonomic benefits of Surface. You'll also notice that we've added an `:on-click` event binding for the `"inc"` and `"dec"` events.  You can get more context for this handle event directive [here](https://surface-ui.org/events#using-the-:on-[event]-directive).
```

Reviewer: I would take the time to explain the event directives a bit more here. This feel like good content to add to an introductory post like this.

And this its how its look like our first suface page:

> ![](./surface-count-live.mov)

Reviewer: It would be great to add a conclusion that wraps up what the user learned and highlight some of the benefits of working with Surface :)
